File: apps/android/app/src/main/java/dev/flextranslate/foundation/GeminiDirectClient.kt
package dev.flextranslate.foundation

import android.util.Log
import org.json.JSONArray
import org.json.JSONObject
import java.io.BufferedReader
import java.net.HttpURLConnection
import java.net.URL

/**
 * BYOK-клиент: шлёт промпт на перевод прямо в публичный REST-эндпоинт Gemini под ключом самого
 * пользователя. Ключ живёт только в заголовке `x-goog-api-key` — нигде не логируется, не попадает
 * в тексты ошибок и не хранится в этом классе.
 *
 * Endpoint: `POST https://generativelanguage.googleapis.com/v1beta/models/{model}:generateContent`
 * Auth:     заголовок `x-goog-api-key: <user key>` (ключ даёт вызывающий, тут не храним)
 *
 * Раскладка ошибок (честная, с поправкой на гео-ограничения):
 *  - HTTP 400 с "User location is not supported" → [DirectResult.GeoBlocked] (показываем как есть)
 *  - HTTP 401 / 403                              → [DirectResult.KeyRejected]
 *  - сбой сети/разбора                            → [DirectResult.Failed]
 *  - успех с непустым текстом                     → [DirectResult.Ok]
 *
 * Безопасность: ключ передаётся аргументом, чтобы вызывающий каждый раз доставал свежее значение
 * из защищённого хранилища; класс не держит ссылку на ключ между вызовами. Ключ уходит только в
 * [setRequestProperty], не захватывается ни одним замыканием и не логируется.
 */
open class GeminiDirectClient(
    private val config: GeminiFlashConfig,
) {

    sealed interface DirectResult {
        data class Ok(val text: String) : DirectResult

        /** HTTP 400 с телом про гео-ограничение — прямой Gemini в этом регионе недоступен. */
        data object GeoBlocked : DirectResult

        /** HTTP 401 или 403 — Google отверг ключ. */
        data object KeyRejected : DirectResult

        /** Сбой транспорта, разбора или неожиданная ошибка сервера. */
        data class Failed(val cause: String) : DirectResult
    }

    /**
     * Переводит [text] для пары [languagePair] прямым вызовом Gemini.
     *
     * @param apiKey Gemini API-ключ пользователя из [GeminiKeyStore]. Вызывающий обязан достать его
     *   из защищённого хранилища прямо перед вызовом, чтобы ключ не висел в памяти дольше нужного.
     *   Не должен быть пустым.
     */
    open fun translate(text: String, languagePair: String, apiKey: String): DirectResult {
        require(apiKey.isNotBlank()) { "apiKey must not be blank" }
        val endpoint = buildEndpoint(config.modelId)
        val body = buildDirectRequestBody(text, languagePair)
        return runCatching { post(endpoint, body, apiKey) }
            .getOrElse { t ->
                // Логируем класс/сообщение исключения — но не ключ и не тело запроса.
                Log.w(TAG, "direct Gemini call failed: ${t.javaClass.simpleName}: ${t.message}")
                DirectResult.Failed(t.javaClass.simpleName)
            }
    }

    private fun post(endpoint: String, body: String, apiKey: String): DirectResult {
        val connection = (URL(endpoint).openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            connectTimeout = config.timeoutMs
            readTimeout = config.timeoutMs
            doOutput = true
            setRequestProperty("Content-Type", "application/json; charset=utf-8")
            setRequestProperty("Accept", "application/json")
            // Ключ пользователя в стандартном заголовке Google AI.
            // Единственное место, где используется значение ключа; не логируется и не сохраняется.
            setRequestProperty("x-goog-api-key", apiKey)
        }
        return try {
            connection.outputStream.use { it.write(body.toByteArray(Charsets.UTF_8)) }
            val status = connection.responseCode
            val stream = if (status in 200..299) connection.inputStream else connection.errorStream
            val payload = stream?.bufferedReader()?.use(BufferedReader::readText).orEmpty()
            parseDirectResponse(status, payload)
        } finally {
            connection.disconnect()
        }
    }

    private companion object {
        const val TAG = "GeminiDirectClient"
        const val GEMINI_BASE = "https://generativelanguage.googleapis.com/v1beta/models"

        fun buildEndpoint(modelId: String): String = "$GEMINI_BASE/$modelId:generateContent"
    }
}

/**
 * Собирает тело запроса Gemini `generateContent` для задачи перевода.
 * В промпте только направление и текст — ни ключа, ни учётных данных.
 */
internal fun buildDirectRequestBody(text: String, languagePair: String): String {
    val (srcCode, tgtCode) = languagePair.split("->").let { parts ->
        Pair(parts.getOrElse(0) { "ru" }, parts.getOrElse(1) { "en" })
    }
    val srcName = languageDisplayName(srcCode)
    val tgtName = languageDisplayName(tgtCode)
    val prompt = "Translate the following $srcName text to $tgtName. " +
        "Reply with only the translation, no explanation:\n$text"

    val part = JSONObject().put("text", prompt)
    val parts = JSONArray().put(part)
    val content = JSONObject().put("parts", parts)
    val contents = JSONArray().put(content)
    return JSONObject().put("contents", contents).toString()
}

/**
 * Разбирает ответ Gemini `generateContent`.
 *
 * Формат успеха: `{ "candidates": [{ "content": { "parts": [{ "text": "..." }] } }] }`
 * Форматы ошибок:
 *  - HTTP 400 с `"User location is not supported"` в сообщении → [GeoBlocked]
 *  - HTTP 401/403 → [KeyRejected]
 *  - всё остальное не-2xx или битое → [Failed]
 */
internal fun parseDirectResponse(status: Int, payload: String): GeminiDirectClient.DirectResult {
    // Сначала ошибки авторизации — ещё до попытки разобрать JSON.
    if (status == 401 || status == 403) return GeminiDirectClient.DirectResult.KeyRejected

    // HTTP 400: либо гео-ограничение, либо кривой запрос.
    if (status == 400) {
        val lowerPayload = payload.lowercase()
        if (lowerPayload.contains("user location is not supported") ||
            lowerPayload.contains("location is not supported")
        ) {
            return GeminiDirectClient.DirectResult.GeoBlocked
        }
        val errorMsg = runCatching {
            JSONObject(payload).optJSONObject("error")?.optString("message") ?: "HTTP 400"
        }.getOrElse { "HTTP 400" }
        return GeminiDirectClient.DirectResult.Failed(errorMsg)
    }

    if (status !in 200..299) {
        return GeminiDirectClient.DirectResult.Failed("HTTP $status")
    }

    val json = runCatching { JSONObject(payload) }.getOrNull()
        ?: return GeminiDirectClient.DirectResult.Failed("unparseable response")

    val text = runCatching {
        json.getJSONArray("candidates")
            .getJSONObject(0)
            .getJSONObject("content")
            .getJSONArray("parts")
            .getJSONObject(0)
            .getString("text")
    }.getOrNull()

    if (text.isNullOrBlank()) {
        return GeminiDirectClient.DirectResult.Failed("empty or missing text in response")
    }
    return GeminiDirectClient.DirectResult.Ok(text.trim())
}

/** Превращает код языка в название для промпта перевода. */
private fun languageDisplayName(code: String): String = when (code) {
    "ru" -> "Russian"
    "en" -> "English"
    "zh" -> "Chinese"
    else -> code
}

File: apps/ios/FlexTranslate/Sources/FlexTranslate/GeminiDirectClient.swift
import Foundation

/// BYOK direct client: POSTs a translate prompt straight to the public Gemini REST endpoint
/// using the user's own API key. The key travels ONLY in the x-goog-api-key request header —
/// it is NEVER logged, NEVER included in error messages, NEVER stored by this class.
///
/// Endpoint: POST https://generativelanguage.googleapis.com/v1beta/models/{model}:generateContent
/// Auth:     x-goog-api-key: <user key> header (supplied by caller, never retained here)
///
/// Error mapping (honest, per geo-restriction reality):
///  - HTTP 400 with "User location is not supported" → .geoBlocked (surfaced honestly)
///  - HTTP 401 / 403                                  → .keyRejected
///  - network / parse failure                         → .failed
///  - success with non-empty text                     → .ok
///
/// Mirrors Android GeminiDirectClient.
class GeminiDirectClient: @unchecked Sendable {

    enum DirectResult: Sendable, Equatable {
        case ok(String)
        /// HTTP 400 with a geo-restriction body — direct Gemini is unavailable in this region.
        case geoBlocked
        /// HTTP 401 or 403 — the key was rejected by Google.
        case keyRejected
        /// Transport, parse, or unexpected server error.
        case failed(String)
    }

    private let config: GeminiFlashConfig
    private static let geminiBase = "https://generativelanguage.googleapis.com/v1beta/models"

    init(config: GeminiFlashConfig) {
        self.config = config
    }

    /// Translate text for languagePair by calling Gemini directly.
    /// apiKey must be the user's key fetched from Keychain immediately before this call.
    func translate(text: String, languagePair: String, apiKey: String) -> DirectResult {
        precondition(!apiKey.isEmpty, "apiKey must not be blank")
        let endpoint = "\(Self.geminiBase)/\(config.modelId):generateContent"
        let body = buildDirectRequestBody(text: text, languagePair: languagePair)
        guard let url = URL(string: endpoint) else {
            return .failed("Invalid Gemini endpoint URL")
        }
        var request = URLRequest(url: url, timeoutInterval: config.timeoutSeconds)
        request.httpMethod = "POST"
        request.setValue("application/json; charset=utf-8", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        // The user's own Gemini API key — the ONLY place the key value is used.
        // It is not logged or stored; the header value is not captured by any closure.
        request.setValue(apiKey, forHTTPHeaderField: "x-goog-api-key")
        request.httpBody = body.data(using: .utf8)

        let semaphore = DispatchSemaphore(value: 0)
        var resultData: Data?
        var httpStatus = 0
        var networkErrorMsg: String?

        URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                networkErrorMsg = error.localizedDescription
            } else if let http = response as? HTTPURLResponse {
                httpStatus = http.statusCode
                resultData = data
            }
            semaphore.signal()
        }.resume()
        semaphore.wait()

        if let msg = networkErrorMsg {
            return .failed(msg)
        }
        let payload = resultData.flatMap { String(data: $0, encoding: .utf8) } ?? ""
        return parseDirectResponse(status: httpStatus, payload: payload)
    }
}

/// Build the Gemini generateContent request body for a translation task.
/// The prompt encodes only the direction and text — no key, no credential.
func buildDirectRequestBody(text: String, languagePair: String) -> String {
    let parts = languagePair.components(separatedBy: "->")
    let srcCode = parts.first ?? "ru"
    let tgtCode = parts.dropFirst().first ?? "en"
    let srcName = geminiLanguageDisplayName(srcCode)
    let tgtName = geminiLanguageDisplayName(tgtCode)
    let prompt = "Translate the following \(srcName) text to \(tgtName). " +
        "Reply with only the translation, no explanation:\n\(text)"

    // Build JSON manually — no Foundation JSONSerialization issues with string escaping
    let escaped = prompt
        .replacingOccurrences(of: "\\", with: "\\\\")
        .replacingOccurrences(of: "\"", with: "\\\"")
        .replacingOccurrences(of: "\n", with: "\\n")
    return "{\"contents\":[{\"parts\":[{\"text\":\"\(escaped)\"}]}]}"
}

/// Parse a Gemini generateContent response.
///
/// Success shape: { "candidates": [{ "content": { "parts": [{ "text": "..." }] } }] }
/// Error shapes:
///  - HTTP 400 with "User location is not supported" → .geoBlocked
///  - HTTP 401/403 → .keyRejected
///  - anything else non-2xx or malformed → .failed
func parseDirectResponse(status: Int, payload: String) -> GeminiDirectClient.DirectResult {
    if status == 401 || status == 403 { return .keyRejected }

    if status == 400 {
        let lower = payload.lowercased()
        if lower.contains("user location is not supported") || lower.contains("location is not supported") {
            return .geoBlocked
        }
        return .failed("HTTP 400: \(payload.prefix(200))")
    }

    guard (200...299).contains(status) else {
        return .failed("HTTP \(status)")
    }

    guard let data = payload.data(using: .utf8),
          let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
          let candidates = json["candidates"] as? [[String: Any]],
          let first = candidates.first,
          let content = first["content"] as? [String: Any],
          let parts = content["parts"] as? [[String: Any]],
          let text = parts.first?["text"] as? String,
          !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    else {
        return .failed("Unparseable or empty Gemini response")
    }
    return .ok(text.trimmingCharacters(in: .whitespacesAndNewlines))
}

private func geminiLanguageDisplayName(_ code: String) -> String {
    switch code {
    case "ru": return "Russian"
    case "en": return "English"
    case "zh": return "Chinese"
    default: return code
    }
}
